File: tps/entrada e saida/tp02/TP02Q13 - Mergesort em Java/Jogador.java
/* Marco Aurelio Silva de Souza Junior - 696809 */

import java.util.*;

public class Jogador {
  static final int maxEntradas = 512;
  static int numComparacoes;
  static int numMovimentacoes;
  static int qtdRegistros; // quantidade de jogadores registrados

  private int id;
  private String nome;
  private int altura;
  private int peso;
  private String universidade;
  private int anoNascimento;
  private String cidadeNascimento;
  private String estadoNascimento;
  
  // Construtores
  public Jogador(int id, String nome, int altura, int peso, String universidade, int anoNascimento, String cidadeNascimento, String estadoNascimento) {
    Jogador.qtdRegistros++;
    setId(id);
    setNome(nome);
    setAltura(altura);
    setPeso(peso);
    setUniversidade(universidade);
    setAnoNascimento(anoNascimento);
    setCidadeNascimento(cidadeNascimento);
    setEstadoNascimento(estadoNascimento);
  }

  // construtor com universidade
  public Jogador(String universidade) {
    Jogador.qtdRegistros++;
    setUniversidade(universidade);
  }

  // construtor recebe array de strings com atributos 
  public Jogador(String atributo[]) {
    Jogador.qtdRegistros++;
    setId(Integer.parseInt(atributo[0]));
    setNome(atributo[1]);
    setAltura(Integer.parseInt(atributo[2]));
    setPeso(Integer.parseInt(atributo[3]));
    setUniversidade(atributo[4]);
    setAnoNascimento(Integer.parseInt(atributo[5]));
    setCidadeNascimento(atributo[6]);
    setEstadoNascimento(atributo[7]);
  }

  // cria clone do objeto
  public Jogador(Jogador clonado) {
    Jogador.qtdRegistros++;
    setId(clonado.id);
    setAltura(clonado.altura);
    setPeso(clonado.peso);
    setUniversidade(clonado.universidade);
    setAnoNascimento(clonado.anoNascimento);
    setCidadeNascimento(clonado.cidadeNascimento);
    setEstadoNascimento(clonado.estadoNascimento);
  }

  //getters n' setters
  public int getId() {
    return this.id; 
  }
  public void setId(int id) {
    this.id = id;
  }

  public String getNome() {
    return this.nome;
  }
  public void setNome(String nome) {
    this.nome = nome;
  }

  public int getAltura() {
    return this.altura;
  }
  public void setAltura(int altura) {
    this.altura = altura;
  }

  public int getPeso() {
    return peso;
  }
  public void setPeso(int peso) {
    this.peso = peso;
  }

  public String getUniversidade() {
    return this.universidade;
  }
  public void setUniversidade(String universidade) {
    if(universidade.isEmpty())
      this.universidade = "ZZZ";
    else 
      this.universidade = universidade;
  }
  
  public int getAnoNascimento() {
    return this.anoNascimento;
  }
  public void setAnoNascimento(int anoNascimento) {
    this.anoNascimento = anoNascimento;
  }

  public String getCidadeNascimento() {
    return this.cidadeNascimento;
  }
  public void setCidadeNascimento(String cidade) {
    this.cidadeNascimento = cidade;
  }

  public String getEstadoNascimento() {
    return this.estadoNascimento;
  }
  public void setEstadoNascimento(String estado) {
    this.estadoNascimento = estado;
  }

  // mostra na tela todos os dados do objeto
  public void imprimir() {
    String vazio = "nao informado";
    MyIO.print("[");
    MyIO.print(getId() + " ## ");
    MyIO.print((getNome().isEmpty() ? vazio : getNome()) + " ## ");
    MyIO.print((getAltura()==-1 ? vazio : getAltura()) + " ## ");
    MyIO.print((getPeso()==-1 ? vazio : getPeso()) + " ## ");
    MyIO.print((getAnoNascimento()==-1 ? vazio : getAnoNascimento()) + " ## ");
    MyIO.print((getUniversidade().equals("ZZZ") ? vazio : getUniversidade()) + " ## ");
    MyIO.print((getCidadeNascimento().isEmpty() ? vazio : getCidadeNascimento()) + " ## ");
    MyIO.println((getEstadoNascimento().isEmpty() ? vazio : getEstadoNascimento()) + "]");
  }

  public static String leLinhaCSV(int id) {
    Arq.openRead("/tmp/players.csv");
    String resp = "";
    // mover cursor do csv ate a linha do id
    for(int n = 0; n <= id + 1; n++) { 
      resp = Arq.readLine();  // le a linha com id inserido e armazena a string em "linha"
    }
    Arq.close();
    return resp;
  }

  // le linha do csv com id indicado e cria um objeto "Jogador"
  public static Jogador ler(int id) {
    String linha = leLinhaCSV(id);  // armazena uma linha de entrada de cada vez 
    String[] atributos = linha.split(",", -1); // divide string de entrada e insere os atributos no array "atributos"
    Jogador j = new Jogador(atributos);
    return j;
  }

  // retorna quantidade de entradas efetuadas
  public static int leituraDeEntrada(int entrada[]) {
    int numEntrada = 0;
    do {
      entrada[numEntrada] = MyIO.readInt();
      numComparacoes++;
    } while (entrada[numEntrada++] != -1);
    numEntrada--;   //Desconsiderar linha contendo a palavra FIM
    return numEntrada;
  }
    
  // retorna quantidade de entradas efetuadas
  public static int leituraDeEntrada(String entrada[]) {
    int numEntrada = 0;
    do {
      entrada[numEntrada] = MyIO.readLine();
      numComparacoes++;
    } while (!entrada[numEntrada++].equals("FIM"));
    numEntrada--;   //Desconsiderar linha contendo a palavra FIM
    return numEntrada;
  }

  // ORDENACAO
  public static void sort(Jogador array[]) {
    int n = array.length;
    mergesort(array, 0, n-1);
 }
 private static void mergesort(Jogador array[], int esq, int dir) {
    if (esq < dir){
       int meio = (esq + dir) / 2;
       mergesort(array, esq, meio);
       mergesort(array, meio + 1, dir);
       intercalar(array, esq, meio, dir);
    }
 }

 public static void intercalar(Jogador array[], int esq, int meio, int dir){
    int n1, n2, i, j, k;

    //Definir tamanho dos dois subarrays
    n1 = meio-esq+1;
    n2 = dir - meio;

    Jogador[] a1 = new Jogador[n1+1];
    Jogador[] a2 = new Jogador[n2+1];

    //Inicializar primeiro subarray
    for(i = 0; i < n1; i++){
      a1[i] = array[esq+i];
      numMovimentacoes++;
    }
    
    //Inicializar segundo subarray
    for(j = 0; j < n2; j++){
      a2[j] = array[meio+j+1];
      numMovimentacoes++;
    }
    
    //Sentinela no final dos dois arrays
    a1[i] = new Jogador("ZZZZZZ");
    a2[j] = new Jogador("ZZZZZZ");

    //Intercalacao propriamente dita
    for(i = j = 0, k = esq; k <= dir; k++) {
      if(
        numComparacoes++ != 0 && 
        a1[i].getUniversidade().compareTo(a2[j].getUniversidade()) < 0 || (
          numComparacoes++ != 0 &&
          a1[i].getUniversidade().compareTo(a2[j].getUniversidade()) == 0 &&
          numComparacoes++ != 0 &&
          a1[i].getNome().compareTo(a2[j].getNome()) < 0
        )
      ) {
        array[k] = a1[i++];
        numMovimentacoes++;
      }
      else {
        array[k] = a2[j++];
        numMovimentacoes++;
      }
    }
 }

  public static void main(String[] args) {
    int[] entrada = new int[maxEntradas]; // array que armazena as linhas de entrada
    
    int numEntrada = leituraDeEntrada(entrada); //Leitura da entrada padrao
    
    Jogador[] jogador = new Jogador[numEntrada]; // criar vetor de objetos Jogador
    
    for(int i = 0; i < numEntrada; i++) {
      jogador[i] = ler(entrada[i]);  // le os dados de "linha" e armazena nos atributos do obj "Jogador"
    }

    long inicio = new Date().getTime();
    sort(jogador);  // ordena vetor pelo nome dos jogadores
    long fim = new Date().getTime();

    for (int i = 0; i < numEntrada; i++) {
      jogador[i].imprimir();  // mostra dados na tela
    }

    // log
    String conteudo = "696809\t" + numComparacoes + "\t" + numMovimentacoes + "\t" + (fim-inicio);
    Arq.openWriteClose("matricula_mergesort.txt", conteudo);
  }
}

File: tps/entrada e saida/tp02/TP02Q08 - Shellsort em C/Jogador.cpp
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <time.h>

#define MAX_ENTRADAS 512  
#define STR_LEN 50   // qtd max de caracteres de uma string
#define TAM_LIN 150  // qtd max de caracteres de uma linha do csv
#define NUM_ATTR 8  // no. de atributos do jogador

int qtdRegistros = 0;
int numComparacoes = 0;
int numMovimentacoes = 0;

// propriedades
typedef struct {
  int id;
  char nome[STR_LEN];
  int altura;
  int peso;
  char universidade[STR_LEN];
  int anoNascimento;
  char cidadeNascimento[STR_LEN];
  char estadoNascimento[STR_LEN];
} Jogador ;

// construtores
void setJogador(char* strAttr[], Jogador* J) {
  J->id = atoi(strAttr[0]);
  strcpy(J->nome, strAttr[1]);
  J->altura = atoi(strAttr[2]);
  J->peso = atoi(strAttr[3]);
  strcpy(J->universidade, strAttr[4]);
  J->anoNascimento = atoi(strAttr[5]);
  strcpy(J->cidadeNascimento, strAttr[6]);
  strcpy(J->estadoNascimento, strAttr[7]);

  qtdRegistros++;
}

void clone(Jogador* clone, Jogador* clonado) {
  *clone = *clonado;
}

// mostra na tela os atributos do Jogador
void imprimir(Jogador* J) {
  char* vazio = "nao informado";
  printf("[");
  printf("%d", J->id); printf(" ## ");
  printf("%s", J->nome[0]=='\0' ? vazio : J->nome); printf(" ## ");
  printf("%d", J->altura); printf(" ## ");
  printf("%d", J->peso); printf(" ## ");
  printf("%d", J->anoNascimento); printf(" ## ");
  printf("%s", J->universidade[0]=='\0' ? vazio : J->universidade); printf(" ## ");
  printf("%s", J->cidadeNascimento[0]=='\0' ? vazio : J->cidadeNascimento); printf(" ## ");
  printf("%s", J->estadoNascimento[0]=='\0' ? vazio : J->estadoNascimento); printf("]\n");
}

// recebe string de uma lihna do csv e coloca os atributos na array
void atribuirAtributos(char* input, char* output[]) {
  int i = 0;
  strtok(input, "\n"); // retira quebra de linha no final da string
  numComparacoes++;
  
  output[i++] = input;
  numMovimentacoes++;

  // loop vai repartindo linha nas ',' e atribuindo a output[]
  while (numComparacoes++ > -1 && i < NUM_ATTR && input && (input = strchr(input, ','))) {
    *input = 0;           
    output[i++] = ++input;
  }
}

// le a lihna do csv correspondente ao id procurado
void lerLinha(FILE* csv, char linha[], int id) {
  rewind(csv);
  for(int n = 0; n <= id + 1; n++) {
    fgets(linha, TAM_LIN, csv);
  }
}

// atribui cada atributo do array de strings no objeto Jogador
Jogador ler(FILE *csv, int id) {
  Jogador J;
  char linha[TAM_LIN]; // linha a ser lida do csv 
  char* strAttr[NUM_ATTR];

  lerLinha(csv, linha, id); 
  
  atribuirAtributos(linha, strAttr);
  
  setJogador(strAttr, &J);
  
  qtdRegistros++;
  return J;
}

// retorna numero de entrada efetuadas
int leituraDeEntrada(char entrada[][STR_LEN]) {
  int numEntrada = 0; // contador de entradas
  do {
    scanf("%[^\n]%*c", entrada[numEntrada]);
  } while(strcmp(entrada[numEntrada++],"FIM")!=0);
  numEntrada--;  // desconsidera entrada com FIM
  return numEntrada;
}

void insercaoPorCor(Jogador *array, int n, int cor, int h){
  for (int i = (h + cor); i < n; i+=h) {
    Jogador tmp = array[i];
    numMovimentacoes++;
    int j = i - h;
    while (
      j >= 0 && ++numComparacoes!=0 && array[j].peso > tmp.peso || (
        array[j].peso == tmp.peso &&
        strcmp(array[j].nome, tmp.nome) > 0
      )
    ) {
      array[j + h] = array[j];
      numMovimentacoes++;
      j-=h;
    }
    array[j + h] = tmp;
    numMovimentacoes++;
  }
}

void shellsort(Jogador *array, int n) {
  int h = 1;

  do { h = (h * 3) + 1; } while (h < n);

  do {
    h /= 3;
    for(int cor = 0; cor < h; cor++){
      insercaoPorCor(array, n, cor, h);
    }
  } while (h != 1);
}

int main(int argc, char *argv[]) {
  Jogador J[MAX_ENTRADAS];
  FILE *csv;
  csv = fopen("/tmp/players.csv", "r");

  char entrada[MAX_ENTRADAS][STR_LEN]; // armazenar entrada padrao com id's
  
  int numEntrada = leituraDeEntrada(entrada); // leitura de entrada padrao

  // percorre as entradas
  for (int i = 0, id; i < numEntrada; i++) {
    id = atoi(entrada[i]);  // recebe valor da entrada padrao para procurar no csv
    J[i] = ler(csv, id);
  }
    
  clock_t inicio = clock();
  shellsort(J, numEntrada);
  clock_t fim = clock();
  
  for (int i = 0; i < numEntrada; i++) {
    imprimir(&J[i]);
  }
 
  fclose(csv);

  // log
  FILE* log = fopen("matricula_shellsort.txt", "w+");
  fprintf(log, "696809\t");
  fprintf(log, "%d\t", numComparacoes);
  fprintf(log, "%d\t", numMovimentacoes);
  fprintf(log, "%ld\t", (fim-inicio));
  fclose(log);

  return 0;
}
